polar_route/crossing_smoothing.py

```diff
@@ -0,0 +1,1437 @@
+            return False
+        
+    def previous_diagonals(self,edge_a,edge_b,firstpoint,lastpoint):
+        '''
+            For a diagonal-additional case determine if we have already seen these edges added in in the 
+            same situation and the same first and last points point. If a common past has been seen return True, 
+            otherwise add an additional case to of the diagonal to the global list and return False
+
+            Input:
+                edge_a (find_edge)     - First-edge connecting start cell to new cell 
+                edge_b (find_edge)     - First-edge connecting new cell  to end cell
+                firstpoint (tuple)     - firstpoint in the adjacent cell tripplet of points (long,lat)
+                lastpoint (tuple)      - lastpoint in the adjacent cell tripplet of points (long,lat)
+
+            Return:
+                True if this U-additional case has been seen before, or false if not
+        
+        '''
+
+        edge_a_start_index = edge_a.start['id']
+        edge_b_start_index = edge_b.start['id']
+        edge_a_end_index   = edge_a.end['id']
+        edge_b_end_index   = edge_b.end['id']
+
+        current_diagonal = [edge_a_start_index,edge_a_end_index,
+                     edge_b_start_index,edge_b_end_index,
+                     firstpoint,lastpoint]
+        
+
+        if len(self.previous_diagonal_info) == 0:
+            self.previous_diagonal_info += [current_diagonal]
+            return False
+
+        if current_diagonal[:-2] in np.array(self.previous_diagonal_info,dtype=object)[:,:-2].tolist():
+            previous_diagonal_info_np = np.array(self.previous_diagonal_info,dtype=object)
+            similar_start_end = previous_diagonal_info_np[(previous_diagonal_info_np[:,:-2] == current_diagonal[:-2]).all(axis=1),-2:]
+            if np.any([self.dist(c[0],current_diagonal[-2]) <= self.merge_separation and self.dist(c[1],current_diagonal[-1]) <= self.merge_separation for c in similar_start_end]):
+                return True
+            else:
+                self.previous_diagonal_info += [current_diagonal]
+                return False
+
+        else:
+            self.previous_diagonal_info += [current_diagonal]
+            return False
+        
+    # def previous_merge(self,firstpoint,midpoint,lastpoint):
+    #     current_merge = [firstpoint,midpoint,lastpoint]
+    #     if len(self.previous_merge_info) == 0:
+    #         self.previous_merge_info += [current_merge]
+    #         return False
+    #     if np.any([self.dist(c[0],current_merge[0]) <= self.converged_sep and 
+    #                self.dist(c[1],current_merge[1]) <= self.converged_sep and 
+    #                self.dist(c[2],current_merge[2]) <= self.converged_sep for c in self.previous_merge_info]):
+    #         return True
+    #     else:
+    #         self.previous_merge_info += [current_merge]
+    #         return False
+        
+    def forward(self):
+        '''
```

Comment:
This is a really long function (_SonarLint gives it a cognitive complexity rating of 106 - it gives warnings at anything over 15_), so debugging or extending it in future may be a real challenge. Could it be broken down into smaller parts?

Reply:
This cannot be changed as its inline with the pseudo code in the paper as outlined by @foxm1